connect/runtime/src/test/java/org/apache/kafka/connect/integration/MonitorableSourceConnector.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.integration;
+
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.Task;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.runtime.TestSourceConnector;
+import org.apache.kafka.connect.source.SourceRecord;
+import org.apache.kafka.connect.source.SourceTask;
+import org.apache.kafka.tools.ThroughputThrottler;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+/**
+ * A source connector to be used in integration tests. The rate of record production can be
```

Comment:
Nit: Are these monitorable connectors really only useful for the Connect integration tests? Seems like they are, and that third party projects won't use these connectors even though they'll create integration tests that use the embedded cluster.

Reply:
You could use such a connector to test your connector end to end. For example if you're building a sink you could use this one to have a pipeline with end-to-end conversions in your integration test (instead of just using a producer). Your concern here is why this isn't in the same package as `EmbeddedConnectCluster` ? 

As a first step I added this class in the same package as `MonitorableSinkConnector` (tbh I would have also named it `MonitoredSourceConnector` but I followed symmetry in naming too).

Reply:
First of all, I think I was thrown off by the first sentence in the JavaDoc ("A source connector _to be used in integration tests_", emphasis mine), which to me sounds like I definitely should use this connector whenever I'm writing an integration test. That's not the case -- I _can_ use it if there's a good reason.

However, I'm having a hard time understanding why I'd want to use it when testing a custom connector, since this connector is pretty restrictive (I can't easily change the structure of the keys or values in the generated `SourceRecord`), harder to control (it's going to operate on it's own timings given the settings), and my integration tests will be relying upon more of the framework to operate correctly. It seems far more straightforward to simply use a producer to write records that a sink connector might consume and consumer to read the records that a source connector produced, and by doing so I'm keeping my test more focused and involving less of the framework (which in my custom connector project I expect to work correctly).

When I wrote my first integration test a while back for a real connector (outside of AK) using the `EmbeddedConnectCluster`, I originally thought I somehow *had* to use the monitorable connectors, `Runtime`, and `ConnectorHandle` instances. But really these are far more useful for testing that the framework is doing the right thing and behaving correctly, than for testing other real connectors. And for testing the framework, "monitorable" connectors are very useful because that's the only way, really, to verify that the framework is working correctly.

So given all this, one suggestion I might have is to change the first line of the JavaDoc to "A source connector that is used in Apache Kafka integration tests to verify the behavior of the Connect framework, but that can be used in other integration tests as a simple connector that generates records of a fixed structure."

Reply:
I definitely agree. I also have used them for framework testing so far too. 
You can always subclass to tune behaviour and this might be handy here. 

But I'm happy to add your javadoc suggestion to show the intent here.

Reply:
yeah, it's cool to have another connector, but for the usecases, would the MonitorableSinkConnector work? I'd imagine to simulate connectors and tasks, that'd be enough. And we would need this additional class only if we were testing something specific to source connectors would we need this class.

Reply:
Done. Btw, since the message for the sink connector was similar, I added the change there too.

Reply:
@wicknicks besides completeness and symmetry which here I find good, source connector class will help when we don't want to interfere with kafka consumer's protocol and also when we want to test source offset commits. I thought hard about adding this, and it's rather useful.